src/machines/java/com/enderio/machines/common/blockentity/task/host/MachineTaskHost.java

```diff
@@ -0,0 +1,149 @@
+import net.minecraftforge.fml.LogicalSide;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.function.Supplier;
+
+public abstract class MachineTaskHost {
+    @Nullable
+    private IMachineTask currentTask;
+
+    private boolean isNewTaskAvailable;
+
+    /**
+     * A serialized task waiting for the level to load.
+     */
+    @Nullable
+    private CompoundTag pendingTask;
+
+    @UseOnly(LogicalSide.CLIENT)
+    private float clientTaskProgress;
+
+    private final Supplier<Level> levelSupplier;
+    private final Supplier<Boolean> canAcceptNewTask;
+
+    /**
+     * This should be constructed in the constructor of your block entity.
+     */
+    public MachineTaskHost(EnderBlockEntity blockEntity, Supplier<Boolean> canAcceptNewTask) {
+        levelSupplier = blockEntity::getLevel;
+        this.canAcceptNewTask = canAcceptNewTask;
+
+        // Add sync data slot for crafting progress
+        blockEntity.addDataSlot(new FloatDataSlot(this::getProgress, p -> clientTaskProgress = p, SyncMode.GUI));
+    }
+
+    @Nullable
+    protected Level getLevel() {
+        return levelSupplier.get();
+    }
+
+    // region Abstract Implementation
+
+    /**
+     * Get the new task.
+     */
+    @Nullable
+    protected abstract IMachineTask getNewTask();
+
+    /**
+     * Load the task from NBT.
+     */
+    @Nullable
+    protected abstract IMachineTask loadTask(CompoundTag nbt);
+
+    // endregion
+
+    // region Task Handling
+
+    @Nullable
+    public IMachineTask getCurrentTask() {
+        return currentTask;
+    }
+
+    public final boolean hasTask() {
+        return currentTask != null;
+    }
+
+    public final float getProgress() {
+        Level level = levelSupplier.get();
+        if (level == null) {
+            return 0;
+        }
+
+        // Client has no knowledge of task, so we use a synced field.
+        if (level.isClientSide) {
+            return clientTaskProgress;
+        }
+
+        if (!hasTask()) {
+            return 0;
+        }
+
+        return currentTask.getProgress();
+    }
+
+    public void tick() {
+        // If we have no active task, get a new one
+        if ((currentTask == null || currentTask.isCompleted()) && isNewTaskAvailable && canAcceptNewTask.get()) {
+            currentTask = getNewTask();
+            isNewTaskAvailable = false;
+        }
+
+        // If we have an unfinished task, continue it.
+        if (currentTask != null && !currentTask.isCompleted()) {
+            currentTask.tick();
+        }
+
+        // If the task finished, next tick we'll try find a new one.
+        // Just in case the task didn't perform an action that caused a task change.
+        if (currentTask != null && currentTask.isCompleted()) {
+            newTaskAvailable();
+        }
+    }
+
+    public final void newTaskAvailable() {
+        isNewTaskAvailable = true;
+    }
+
+    public final void onLevelReady() {
+        // Load any pending tasks.
```

Comment:
This is still the same `OnLoad` logic, right? So it still doesn't run properly when placing a block (needed for the spawner).

Reply:
Yeah it is, whats not working exactly?

Reply:
that OnLoad is called before Load when placing a block, making it so the BlockEntityData with the entitytype is empty when the task is created. After adding power the task gets remade and is fine.